### DISCUSSION TITLE
How to save an array of values related to a model to the database ?

### DISCUSSION THREAD

**Discussion by ghost** [adonisjs/core#3311] [Help] [answered]
Hi,
1. I have `featured_products` table and `featured_products_items` table.
2. Each featured product **has many** featured products items.
3. I created the models of both:

`app/Models/FeaturedProduct.ts`
```ts
import { DateTime } from "luxon";
import {
  BaseModel,
  column,
  HasMany,
  hasMany,
  HasOne,
  hasOne,
} from "@ioc:Adonis/Lucid/Orm";
import FeaturedProductImage from "./FeaturedProductImage";
import FeaturedProductItem from "./FeaturedProductItem";

export default class FeaturedProduct extends BaseModel {
  @column({ isPrimary: true })
  public id: number;

  @column()
  public productId: number;

  @column()
  public label: string;

  @column()
  public title: string;

  @column.dateTime({ autoCreate: true })
  public createdAt: DateTime;

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  public updatedAt: DateTime;

  @hasOne(() => FeaturedProductImage)
  public image: HasOne<typeof FeaturedProductImage>;

  @hasMany(() => FeaturedProductItem)
  public items: HasMany<typeof FeaturedProductItem>;
}
```

`app/Models/FeaturedProductItem.ts`
```ts
import { DateTime } from "luxon";
import { BaseModel, column } from "@ioc:Adonis/Lucid/Orm";

export default class FeaturedProductItem extends BaseModel {
  @column({ isPrimary: true })
  public id: number;

  @column()
  public featuredProductId: number;

  @column()
  public item: string;

  @column.dateTime({ autoCreate: true })
  public createdAt: DateTime;

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  public updatedAt: DateTime;
}
```

Now in the `app/Controllers/Http/FeaturedProductsController.ts` i wanna store the featured product. The incoming data from the user is as follow:
```json
{
    "product_id": "1",
    "label": "{{$randomWord}}",
    "title": "{{$randomWord}}",
    "items": [
        "{{$randomWord}} {{$randomWord}} {{$randomWord}}",
        "{{$randomWord}} {{$randomWord}} {{$randomWord}}",
        "{{$randomWord}} {{$randomWord}} {{$randomWord}}"
    ]
}
```

The store method in `app/Controllers/Http/FeaturedProductsController.ts` is as follow:
```ts
public async store({ request, bouncer }: HttpContextContract) {
    const payload = await request.validate(CreateFeaturedProductValidator);
    await bouncer.with("FeaturedProductPolicy").authorize("store");
    const featuredProduct = new FeaturedProduct();
    featuredProduct.productId = payload.product_id;
    featuredProduct.label = payload.label;
    featuredProduct.title = payload.title;
    await featuredProduct.save();
    
    // This fails 
    await featuredProduct.related("items").save(payload.items);
    
    // also this fails
    for (const item of payload.items) {
      await featuredProduct.related("items").save();
    }
    return { featuredProduct };
  }
```

Now i am getting these error and i cannot save the items to the database:
![Screenshot from 2021-11-06 16-31-50](https://user-images.githubusercontent.com/76194919/140611563-3951714b-5ffd-4db1-b154-b7b7bffe2bfa.png)
![Screenshot from 2021-11-06 16-45-41](https://user-images.githubusercontent.com/76194919/140611898-7b13d174-d191-4938-8652-2e63121cfbd9.png)

**Comment by tomgobich** [comment, marked as answer]
Hi! The error you're getting in VSCode is because `save` is expecting an actual instance of `FeaturedProductItem`, see https://docs.adonisjs.com/reference/orm/relations/has-many#save. So to use save you'd want to do something like:
```typescript
const featuredProductItems = payload.items.map(item => {
  const featuredProductItem = new FeaturedProductItem()
  featuredProductItem.item = item
  return featuredProductItem
})

await featuredProduct.related('items').saveMany(featuredProductItems)
```

I think in this case, based on what you have written, using `create` or `createMany` is more along the lines of what you want. Unlike `save` and `saveMany`, `create` and `createMany` don't expect an instance of the model, but rather just an object of its properties. So you could do:

```typescript
const featuredProductItems = payload.items.map(item => ({ item }))
await featuredProduct.related('items').createMany(featuredProductItems)
```

or, if you prefer one at a time:
```typescript
for (const item of payload.items) {
  await featuredProduct.related('items').create({ item })
}
```

Here's the documentation for create and createMany:
https://docs.adonisjs.com/reference/orm/relations/has-many#create

Hope this helps!!

**Reply by ghost** [reply]
Please make a lecture in the future where you explain this instance of a model thing and relationships in depth. I read the docs multiple times but could not do it. I think maybe cuz i didn't study typescript before diving into adonis.

Thanks for help that worked.

**Reply by tomgobich** [reply]
Awesome, glad to hear that helped!! I intend on spending many lessons specifically on CRUD and relationships! :)
